### PR TITLE
dag/expr: implement case_when

### PR DESCRIPTION
**Pull request by AndreMouche**
@BusyJay @hicqu @XuHuaiyu PTAL

### CODE REVIEW

File: src/coprocessor/dag/expr/builtin_control.rs [hunk 1]
```diff
@@ -44,6 +44,30 @@ where
...
 }
 
+/// See https://dev.mysql.com/doc/refman/5.7/en/case.html
+fn case_when<F, T>(expr: &FnCall, ctx: &StatementContext, row: &[Datum], f: F) -> Result<Option<T>>
+where
+    F: Fn(usize) -> Result<Option<T>>,
+{
+    let len = expr.children.len();
+    let pairs = len >> 1;
+    for id in 0..pairs {
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/builtin_control.rs]: You mean `for chunk in expr.children.chunks(2)`?

Comment (BusyJay) [reply src/coprocessor/dag/expr/builtin_control.rs]: Ref: https://github.com/pingcap/tikv/blob/master/src/coprocessor/select/xeval/evaluator.rs#L388

File: src/coprocessor/dag/expr/fncall.rs [hunk 2]
```diff
@@ -168,6 +169,16 @@ impl FnCall {
...
             ScalarFuncSig::IfDuration => (3, 3),
+
+            ScalarFuncSig::CaseWhenDecimal |
+            ScalarFuncSig::CaseWhenDuration |
+            ScalarFuncSig::CaseWhenInt |
+            ScalarFuncSig::CaseWhenJson |
+            ScalarFuncSig::CaseWhenReal |
+            ScalarFuncSig::CaseWhenString |
+            ScalarFuncSig::CaseWhenTime => (1, usize::MAX),
+            _ => return Err(box_err!("TODO")),
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/fncall.rs]: Return unknown signature instead.

**Comment by hicqu** [comment]
LGTM.

**Pull request merged**